File: src/606.Construct-String-from-Binary-Tree/solution.go
package solution

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
import "strconv"

func tree2str(root *TreeNode) string {
    if root == nil {
        return ""
    }
    s := strconv.Itoa(root.Val)
    l := tree2str(root.Left)
    r := tree2str(root.Right)

    // Special case1: s()() -> s
    if root.Left == nil && root.Right == nil {
        return s
    }
    // Special case2: s(l)() -> s(l)
    if root.Left != nil && root.Right == nil {
        return s + "(" + l + ")"
    }
    // General case: s(l)(r)
    return s + "(" + l + ")" + "(" + r + ")"
}

func tree2str2(root *TreeNode) string {
    // Rucursion stop case.
    if root == nil {
        return ""
    }
    // Special case1: root()() -> root
    if root.Left == nil && root.Right == nil {
        return strconv.Itoa(root.Val)
    }
    // Special case2: root(l)() -> root(l)
    if root.Right == nil {
        return strconv.Itoa(root.Val) + "(" + tree2str(root.Left) + ")"
    }
    // General case: root(l)(r)
    return strconv.Itoa(root.Val) + "(" + tree2str(root.Left) + ")" + "(" + tree2str(root.Right) + ")"
}

File: src/606.Construct-String-from-Binary-Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string tree2str(TreeNode* root) {
        if (root == nullptr) return "";
        const string s = std::to_string(root->val);
        const string l = tree2str(root->left);
        const string r = tree2str(root->right);

        // Special case1: left + right node == nullptr -> s
        if (root->left == nullptr && root->right == nullptr) return s;
        // Special case2: right == nullptr; left != nullptr -> s(l)
        if (root->left != nullptr && root->right == nullptr) return s+"("+l+")";
        // General case: s(l)(r)
        return s+"("+l+")"+"("+r+")";
    }
};